x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java

```diff
@@ -341,24 +405,34 @@ protected void doSaveState(IndexerState indexerState, Map<String, Object> positi
                 return;
             }
 
-            if(indexerState.equals(IndexerState.STARTED)) {
-                // if the indexer resets the state to started, it means it is done, so increment the generation
+            if(indexerState.equals(IndexerState.STARTED) && getStats().getNumDocuments() > 0) {
+                // if the indexer resets the state to started, it means it is done with a run through the data.
+                // But, if there were no documents, we should allow it to attempt to gather more again, as there is no risk of overwriting
+                // Some reasons for no documents are (but is not limited to):
+                // * Could have failed early on search or index
+                // * Have an empty index
+                // * Have a query that returns no documents
                 generation.compareAndSet(0L, 1L);
             }
 
-            final DataFrameTransformState state = new DataFrameTransformState(indexerState, getPosition(), generation.get());
+            final DataFrameTransformState state = new DataFrameTransformState(
+                taskState.get(),
+                indexerState,
+                getPosition(),
+                generation.get(),
+                stateReason.get());
             logger.info("Updating persistent state of transform [" + transform.getId() + "] to [" + state.toString() + "]");
-
-            updatePersistentTaskState(state, ActionListener.wrap(task -> next.run(), exc -> {
-                logger.error("Updating persistent state of transform [" + transform.getId() + "] failed", exc);
-                next.run();
-            }));
+            persistStateToClusterState(state, ActionListener.wrap(t -> next.run(), e -> next.run()));
         }
 
         @Override
         protected void onFailure(Exception exc) {
-            auditor.error(transform.getId(), "Data frame transform failed with an exception: " + exc.getMessage());
-            logger.warn("Data frame transform [" + transform.getId() + "] failed with an exception: ", exc);
+            if (exc.getMessage().equals(lastAuditedExceptionMessage) == false) {
```

Comment:
could you add a comment what this is doing?

This might solve a problem that has a different root cause: I guess this is about short firing repeated failure messages, but as written above the problem is the scheduling logic.

It's good to have this for now, but we should revisit once we fix the scheduler.